ouroboros-network/test/Test/Mux.hs

```diff
@@ -390,61 +517,275 @@ prop_mux_starvation response0 response1 =
-                      -> Property
-prop_mux_short_header badSdu shortCutOffArg = ioProperty $ do
-    let shortCutOff = fromIntegral $ min shortCutOffArg 7 -- XXX Quickcheck 'Giving up' hack
-    (client_w, serverAsync) <- failingServer
-
-    atomically $ writeTBQueue client_w (BL.take shortCutOff $ encodeInvalidMuxSDU badSdu)
-
-    res <- wait serverAsync
-    case res of
-         Just e  ->
-             case fromException e of
-                  Just me -> return $ Mx.errorType me == Mx.MuxDecodeError
-                  Nothing -> return False
-         Nothing -> return False
-
--- | Start a ReqResp server that is expected to fail, returns the server's read channel
-failingServer :: IO (TBQueue IO BL.ByteString, Async IO (Maybe SomeException))
-failingServer  = do
-    server_r <- atomically $ newTBQueue 10
-    server_w <- atomically $ newTBQueue 10
-    endMpsVar <- atomically $ newTVar 2
+-- | Verify ingress processing of valid and invalid SDUs.
+prop_demux_sdu :: forall m.
+                    ( MonadAsync m
+                    , MonadCatch m
+                    , MonadFork m
+                    , MonadMask m
+                    , MonadSay m
+                    , MonadST m
+                    , MonadSTM m
+                    , MonadThrow (STM m)
+                    , MonadTime m
+                    , MonadTimer m)
+                 => ArbitrarySDU
+                 -> m Property
+prop_demux_sdu a = do
+    r <- run a
+    return $ tabulate "SDU type" [stateLabel a] $
+             tabulate "SDU Violation " [violationLabel a] r
+  where
+    run (ArbitraryValidSDU sdu state (Just Mx.MuxIngressQueueOverRun)) = do
+        stopVar <- newEmptyTMVarM
 
-    let sduLen = 1260
-        request  = DummyPayload $ BL.replicate 4 0xa
-        response = request
+        let server_mp = Mx.MiniProtocolDescription Nothing (Just $ serverRsp stopVar)
+        -- To trigger MuxIngressQueueOverRun we use a special test protocol with
+        -- an ingress queue which is less than 0xffff so that it can be triggered by a
+        -- single segment.
+        let server_small ChainSyncSmall = server_mp
+
+        (client_w, said) <- plainServer server_small
+        setup state client_w
+
+        writeSdu client_w $! unDummyPayload sdu
+
+        atomically $! putTMVar stopVar $ unDummyPayload sdu
 
-    (_, _, server_mp) <- setupMiniReqRsp (return ()) endMpsVar request response
-    let server_mps ChainSync1 = server_mp
+        res <- wait said
+        case res of
+            Just e  ->
+                case fromException e of
+                    Just me -> return $ Mx.errorType me === Mx.MuxIngressQueueOverRun
+                    Nothing -> return $ property False
+            Nothing -> return $ property False
+
+    run (ArbitraryValidSDU sdu state err_m) = do
+        stopVar <- newEmptyTMVarM
+
+        let server_mp = Mx.MiniProtocolDescription Nothing (Just $ serverRsp stopVar)
+        let server_std ChainSync1 = server_mp
```

Comment:
I think you can just inline the `server_mp` in `server_std`.

Reply:
Will address.